OPAL/si/src/main/scala/org/opalj/fpcf/AnalysisScenario.scala

```diff
@@ -225,66 +232,861 @@ class AnalysisScenario[A](val ps: PropertyStore) {
+            // Each strategy defines how computations are scheduled into batches based on their dependencies and other constraints.
+            scheduleStrategy match {
+                case "SPS" =>
+                    // SPS (Single Phase Scheduling) schedules all computations in a single batch.
+                    this.scheduleBatches = List(computePhase(ps, allCS, Set.empty))
+
+                case "MPS" =>
+                    // MPS (Maximum Phase Scheduling) breaks down computations into multiple phases based on dependencies and computation types.
+
+                    // Creates a map from ComputationSpecifications to their indices
+                    val computationSpecificationMap: Map[ComputationSpecification[A], Int] = allCS.zipWithIndex.toMap
+                    // Initializes an empty schedule graph
+                    var scheduleGraph: Map[Int, Set[Int]] = Map.empty
+
+                    // Helper functions and logic for MPS scheduling follow...
+                    def getAllCSFromPropertyBounds(
+                        properties: Set[PropertyBounds]
+                    ): Set[ComputationSpecification[A]] = {
+                        def containsProperty(cs: ComputationSpecification[A], property: PropertyBounds): Boolean =
+                            cs.derivesLazily.contains(property) ||
+                                cs.derivesCollaboratively.contains(property) ||
+                                cs.derivesEagerly.contains(property)
+
+                        allCS.filter(cs => properties.exists(containsProperty(cs, _)))
+                    }
+
+                    def mapCSToNum(specifications: Set[ComputationSpecification[A]]): Set[Int] = {
+                        specifications.flatMap(computationSpecificationMap.get)
+                    }
+
+                    def edgeFunctionForSCCS(node: Int): IntIterator = {
+                        val edges = scheduleGraph.getOrElse(node, Set.empty).iterator
+                        new IntIterator {
+                            def hasNext: Boolean = edges.hasNext
+                            def next(): Int = edges.next()
+                        }
+                    }
+
+                    def getAllUses(css: List[Int]): Set[PropertyBounds] = {
+                        var allUses: Set[PropertyBounds] = Set.empty
+                        css.foreach { cs =>
+                            allUses = allUses ++ computationSpecificationMap.find(_._2 == cs).map(_._1).head.uses(ps)
+                        }
+                        allUses
+                    }
+
+                    def setLazyInAllBatches(
+                        map:          Map[List[Int], Set[Int]],
+                        firstElement: List[Int]
+                    ): Map[List[Int], Set[Int]] = {
+                        var visited_batches: List[List[Int]] = List.empty
+                        var aCyclicGraph = map.toMap
+
+                        def setLazyInAllBatches_rek(
```

Comment:
Naming covention would be to use `XYInternal` or `XYInner` or something more descriptive if possible. Are you even sure this is needed an cannot be replaced with something like a fold, map, or foreach?